File: problems/回溯/93.复原IP地址.py
#!/usr/bin/env python3
"""
93. 复原 IP 地址 (Restore IP Addresses)
题目: 给定一个只包含数字的字符串 s，向其中插入点号来形成有效的 IP 地址。
算法: 回溯
时间复杂度: O(3^4) = O(1)
空间复杂度: O(1)
"""

def restoreIpAddresses(s):
    result = []
    path = []
    
    def is_valid(segment):
        if len(segment) > 1 and segment[0] == '0':
            return False
        return 0 <= int(segment) <= 255
    
    def backtrack(start):
        if len(path) == 4:
            if start == len(s):
                result.append('.'.join(path))
            return
        for i in range(start, min(start + 3, len(s))):
            segment = s[start:i + 1]
            if is_valid(segment):
                path.append(segment)
                backtrack(i + 1)
                path.pop()
    
    backtrack(0)
    return result

if __name__ == "__main__":
    # 测试用例1
    print(restoreIpAddresses("25525511135"))
    # 输出: ["255.255.11.135", "255.255.111.35"]
    
    # 测试用例2
    print(restoreIpAddresses("0000"))
    # 输出: ["0.0.0.0"]
    
    # 测试用例3
    print(restoreIpAddresses("101023"))
    # 输出: ["1.0.10.23", "1.0.102.3", "10.1.0.23", "10.10.2.3", "101.0.2.3"]

File: problems/回溯/93.复原IP地址.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

bool isValid(const string& s, int start, int end) {
    if (end - start + 1 > 3) return false;
    if (end - start + 1 > 1 && s[start] == '0') return false;
    int num = 0;
    for (int i = start; i <= end; i++) {
        num = num * 10 + (s[i] - '0');
        if (num > 255) return false;
    }
    return true;
}

void backtrack(vector<string>& result, string& path,
               const string& s, int start, int parts) {
    if (parts == 4) {
        if (start == s.size()) {
            result.push_back(path);
        }
        return;
    }
    for (int i = start; i < s.size() && i < start + 3; i++) {
        if (isValid(s, start, i)) {
            string segment = s.substr(start, i - start + 1);
            string newPath = path.empty() ? segment : path + "." + segment;
            backtrack(result, newPath, s, i + 1, parts + 1);
        }
    }
}

vector<string> restoreIpAddresses(const string& s) {
    vector<string> result;
    string path;
    if (s.size() >= 4 && s.size() <= 12) {
        backtrack(result, path, s, 0, 1);
    }
    return result;
}

int main() {
    // 测试用例1
    vector<string> res1 = restoreIpAddresses("25525511135");
    cout << "Test 1: \"25525511135\"" << endl;
    for (auto& ip : res1) {
        cout << ip << " ";
    }
    cout << endl;
    
    // 测试用例2
    vector<string> res2 = restoreIpAddresses("0000");
    cout << "Test 2: \"0000\"" << endl;
    for (auto& ip : res2) {
        cout << ip << " ";
    }
    cout << endl;
    
    // 测试用例3
    vector<string> res3 = restoreIpAddresses("101023");
    cout << "Test 3: \"101023\" - " << res3.size() << " addresses" << endl;
    for (auto& ip : res3) {
        cout << ip << " ";
    }
    cout << endl;
    
    return 0;
}
